### PR TITLE
watermark: remove duplicated code / variables

### PR DESCRIPTION
**Pull request by phweyland**
This is the first step. The next step will be to complete variables
with watermark usefull tags (HOUR_AMPM, ...)

To ensure backward compatibility I've added the watermark keywords into variables. 
What is the deprecation process ? Just announce in dtdocs that these keywords will disappear with the next dt version ?

### CODE REVIEW

File: src/common/variables.c [hunk 1]
```diff
@@ -466,7 +466,7 @@ static char *get_base_value(dt_variables_params_t *params, char **variable)
     }
     g_list_free_full(res, &g_free);
   }
-  else if(has_prefix(variable, "DESCRIPTION"))
+  else if(has_prefix(variable, "DESCRIPTION") || has_prefix(variable, "vXmp.dc.description"))
```
Comment (TurboGit) [inline src/common/variables.c:469]: vXmp -> Xmp

File: src/common/variables.c [hunk 2]
```diff
@@ -213,32 +213,32 @@ static char *get_base_value(dt_variables_params_t *params, char **variable)
...
+  else if(has_prefix(variable, "EXIF_DAY") || has_prefix(variable, "EXIF.DATE.DAY"))
     result = g_strdup_printf("%.2d", datetime.day);
-  else if(has_prefix(variable, "EXIF_HOUR"))
+  else if(has_prefix(variable, "EXIF_HOUR") || has_prefix(variable, "EXIF.DATE.HOUR"))
     result = g_strdup_printf("%.2d", datetime.hour);
-  else if(has_prefix(variable, "EXIF_MINUTE"))
+  else if(has_prefix(variable, "EXIF_MINUTE") || has_prefix(variable, "EXIF.DATE.MINUTE"))
     result = g_strdup_printf("%.2d", datetime.minute);
-  else if(has_prefix(variable, "EXIF_SECOND"))
+  else if(has_prefix(variable, "EXIF_SECOND") || has_prefix(variable, "EEXIF.DATE.SECOND"))
```
Comment (TurboGit) [inline src/common/variables.c:241]: EEXIF -> EXIF

**Comment by TurboGit** [comment]
> What is the deprecation process ? Just announce in dtdocs that these keywords will disappear with the next dt version ?

Do we really need to deprecate? Couldn't we keep all aliases?

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some typos.

**Comment by TurboGit** [comment]
> Just announce in dtdocs that these keywords will disappear with the next dt version ?

I even think that we don't (can't) want to do that as it will break any edits using watermarks with those old variables.

**Comment by phweyland** [comment]
> I even think that we don't (can't) want to do that as it will break any edits using watermarks with those old variables.

There are some deprecated modules ... But no issue, and in any case, no hurry. For the time being we keep the compatibility.
Thanks for the typos. I'll fix that.

File: src/common/variables.c [hunk 3]
```diff
@@ -187,71 +190,144 @@ static void init_expansion(dt_variables_params_t *params, gboolean iterate)
...
+    result[3] = '\0';
+  }
+  // for watermark backward compatibility
+  else if(_has_prefix(variable, "DATE"))
+  {
+    dt_datetime_gdatetime_to_exif(exif_datetime, params->data->show_msec ? DT_DATETIME_LENGTH : DT_DATETIME_EXIF_LENGTH, params->data->time);
+    result = g_strdup(exif_datetime);
+  }
+
+  else if(_has_prefix(variable, "EXIF_YEAR") || _has_prefix(variable, "EXIF.DATE.LONG_YEAR"))
```
Comment (TurboGit) [inline src/common/variables.c:290]: Likewise here `LONG_YEAR` better to be `YEAR.LONG`, and reads better to me `EXIF.DATE.YEAR.LONG`.

File: src/common/variables.c [hunk 4]
```diff
@@ -262,49 +338,39 @@ static char *get_base_value(dt_variables_params_t *params, char **variable)
       result = res;
     }
   }
-  else if(has_prefix(variable, "EXIF_APERTURE"))
+  else if(_has_prefix(variable, "EXIF_APERTURE"))
```
Comment (TurboGit) [inline src/common/variables.c:341]: `EXIF_APERTURE` -> `EXIF.APERTURE`

For compatibility we want to keep old namings.

File: src/common/variables.c [hunk 5]
```diff
@@ -423,8 +494,8 @@ static char *get_base_value(dt_variables_params_t *params, char **variable)
         break;
     }
   }
-  else if((has_prefix(variable, "LABELS_ICONS") ||
-           has_prefix(variable, "LABELS_COLORICONS"))
+  else if((_has_prefix(variable, "LABELS_ICONS") ||
```
Comment (TurboGit) [inline src/common/variables.c:497]: likewise : `LABELS.ICONS` & `LABELS.COLORICONS`.

Comment (phweyland) [reply src/common/variables.c:497]: Do you mean that all '_' should become '.' ?

Comment (TurboGit) [reply src/common/variables.c:497]: Yes!

Comment (phweyland) [reply src/common/variables.c]: Done. I haven't included DATE trying to keep things as short as possible.

Comment (phweyland) [reply src/common/variables.c]: I've also grouped MAKER, MODEL and LENS under EXIF, but that may be questionable.

File: src/gui/gtkentry.c [hunk 6]
```diff
@@ -169,78 +169,78 @@ void dt_gtkentry_setup_completion(GtkEntry *entry, const dt_gtkentry_completion_
...
-          { "FILE_NAME", N_("$(FILE_NAME) - basename of the input image") },
-          { "FILE_EXTENSION", N_("$(FILE_EXTENSION) - extension of the input image") },
+      = { { "ROLL.NAME", N_("$(ROLL.NAME) - roll of the input image") },
+          { "FILE.FOLDER", N_("$(FILE.FOLDER) - folder containing the input image") },
+          { "FILE.NAME", N_("$(FILE.NAME) - basename of the input image") },
+          { "FILE.EXTENSION", N_("$(FILE.EXTENSION) - extension of the input image") },
           { "VERSION", N_("$(VERSION) - duplicate version") },
-          { "VERSION_IF_MULTI", N_("$(VERSION_IF_MULTI) - same as $(VERSION) but null string if only one version exists") },
-          { "VERSION_NAME", N_("$(VERSION_NAME) - version name from metadata") },
+          { "VERSION.IF.MULTI", N_("$(VERSION.IF.MULTI) - same as $(VERSION) but null string if only one version exists") },
```
Comment (TurboGit) [inline src/gui/gtkentry.c:177]: Just a quick review, this should be `VERSION.IF_MULTI` I think as IF MULTI are not classification but an single word to me.

Comment (phweyland) [reply src/gui/gtkentry.c:177]: Right!

Comment (phweyland) [reply src/gui/gtkentry.c]: done

File: src/common/variables.c [hunk 7]
```diff
@@ -187,71 +190,144 @@ static void init_expansion(dt_variables_params_t *params, gboolean iterate)
...
-    result = g_strdup_printf("%.2d", datetime.second);
-  else if(has_prefix(variable, "EXIF_MSEC"))
-    result = g_strdup_printf("%.3d", datetime.msec);
-  else if(has_prefix(variable, "EXIF_ISO"))
+  char exif_datetime[DT_DATETIME_LENGTH];
+  GDateTime *datetime = params->data->have_exif_dt ? params->data->datetime : params->data->time;
+
+  if(_has_prefix(variable, "YEAR.SHORT") || _has_prefix(variable, "SHORT_YEAR") || _has_prefix(variable, "DATE.SHORT_YEAR"))
+    result = g_date_time_format(params->data->time, "%y");
+  else if(_has_prefix(variable, "YEAR") || _has_prefix(variable, "DATE.LONG_YEAR"))
```
Comment (TurboGit) [inline src/common/variables.c:260]: Don't we want YEAR.LONG here as synonym?

File: src/gui/gtkentry.c [hunk 8]
```diff
@@ -169,69 +169,78 @@ void dt_gtkentry_setup_completion(GtkEntry *entry, const dt_gtkentry_completion_
...
+          { "WIDTH.RAW", N_("$(WIDTH.RAW) - RAW image width") },
+          { "WIDTH.CROP", N_("$(WIDTH.CROP) - image width after crop") },
+          { "WIDTH.EXPORT", N_("$(WIDTH.EXPORT) - exported image width") },
+          { "HEIGHT.MAX", N_("$(HEIGHT.MAX) - maximum image export height") },
+          { "HEIGHT.SENSOR", N_("$(HEIGHT.SENSOR) - image sensor height") },
+          { "HEIGHT.RAW", N_("$(HEIGHT.RAW) - RAW image height") },
+          { "HEIGHT.CROP", N_("$(HEIGHT.CROP) - image height after crop") },
+          { "HEIGHT.EXPORT", N_("$(HEIGHT.EXPORT) - exported image height") },
           { "YEAR", N_("$(YEAR) - year") },
+          { "YEAR.SHORT", N_("$(YEAR.SHORT) - year without century") },
```
Comment (TurboGit) [inline src/gui/gtkentry.c:192]: And YEAR.LONG here too?

**Review by TurboGit — COMMENTED** [review]
Just a comment while reviewing this again.

File: src/gui/gtkentry.c (hunk 8 above)
Comment (phweyland) [reply src/gui/gtkentry.c:192]: Idem, but I'll would be yet more cautious here, as that's the visible dictionary. So far we don't have synonym. If we start, there is more difficult not to put also all the others.

Comment (TurboGit) [reply src/gui/gtkentry.c:192]: Ok, let's this as-is at least for now. Thanks @phweyland for all the work you put on this.

**Review by TurboGit — APPROVED** [review]
Good to go!

**Pull request merged**